resolwe/storage/connectors/registry.py

```diff
@@ -0,0 +1,92 @@
+    __init_subclass__ in the BaseStorageConnector class when connector classes
+    are imported.
+
+    This class is implemented as singleton: constructing multiple instances
+    of it will always return the same instance.
+    """
+
+    __instance = None  #  A single instance of this class
+
+    @classmethod
+    def __new__(cls, *args, **kwargs):
+        if StorageConnectors.__instance is None:
+            StorageConnectors.__instance = super().__new__(cls)
+            StorageConnectors.__instance.__setattr__("_connectors", dict())
+        return StorageConnectors.__instance
+
+    def _get_connectors_settings(self):
+        """Return settings for storage connectors. Since this module can be
+        imported from inside executor special care should be taken from where
+        settings are read.
+        """
+        key = "STORAGE_CONNECTORS"
+        try:
+            from django.conf import settings
+            settings = getattr(settings, key) if hasattr(settings, key) else {}
+        except ImportError:
+            from executors.global_settings import SETTINGS as settings
+            settings = settings.get(key, {})
+        return settings
+
+    def add_storage_connector_class(self, connector: BaseStorageConnector):
+        """Add storage connector class to the registry. When there are
+        configuration entries in the settings for this class then
+        corresponding instances are created and stored in the registry.
+
+        If connector in not Class then exception is raised.
+        """
+
+        def _get_class_name(klass):
+            return "{}.{}".format(klass.__module__, klass.__name__)
+
+        assert inspect.isclass(connector)
+        assert issubclass(connector, BaseStorageConnector)
+        full_class_name = _get_class_name(connector)
+        logging.info("Adding connector {} to register".format(full_class_name))
+        self._add_connectors_from_settings(full_class_name, connector)
+
+    def _add_connectors_from_settings(self, full_class_name, klass):
+        """Iterate through connectors defined in the settings for the given
+        class name, create their instances and add it to the registry if a
+        match is found.
+        """
+        connectors_settings = self._get_connectors_settings()
+        for connector_name, connector_settings in connectors_settings.items():
+            if connector_settings["connector"] == full_class_name:
+                instance = klass(connector_settings["config"], connector_name)
+                self[connector_name] = instance
+
+    def __getitem__(self, item):
+        return self._connectors.get(item)
+
+    def __iter__(self):
+        return iter(self._connectors)
+
+    def __contains__(self, name):
+        return name in self._connectors
+
+    def __len__(self):
+        return len(self._connectors)
+
+    def __setitem__(self, name, connector):
+        assert name not in self
```

Comment:
It is actually equivalent since self can be used as a map and __contains__ looks in self._connectors. Anyway I have removed this statement since I wanted to recreate connectors when settings change (when running tests for example).